qmt/physics_constants.py

```diff
@@ -127,4 +113,48 @@ def to_float(expr):
 matrices.tau_zy = kron(matrices.s_z, matrices.s_y)
 matrices.tau_zz = kron(matrices.s_z, matrices.s_z)
 
-__all__ = ["units", "constants", "matrices", "parse_unit", "to_float"]
+
+class ArrayQuantity(np.ndarray):
+    """
+    Extend a numpy array to have units information from sympy
+    From https://docs.scipy.org/doc/numpy/user/basics.subclassing.html#simple-example-adding-an-extra-attribute-to-ndarray
+    """
+    def __new__(cls, input_array, unit=None):
+        # Input array is an already formed ndarray instance
+        # We first cast to be our class type
+        obj = np.asarray(input_array).view(cls)
+        # add the unit to the created instance
+        obj.unit = unit
+        obj.dtype = input_array.dtype
+        # Finally, we must return the newly created object:
+        return obj
+
+    def __array_finalize__(self, obj):
+        # see InfoArray.__array_finalize__ for comments
+        if obj is None:
+            return
+        self.info = getattr(obj, 'info', None)
+
+    def is_dimensionless(self):
+        return self.unit == None
+
+
+class Quantity(float):
+    """
+    A simple subclass of float that includes a unit
+    Note that this is only a wrapper that does not support arithematic operations or comparisons.
```

Comment:
Removing Quantity and ArrayQuantity in favor of float/np array multiplied by a sympy unit

Reply:
@likewei92 - Sorry for chiming in late, but do I understand correctly that you propose to store, say, potential data as an `ndarray(dtype=object)` of sympy expressions, which is what you get from `potential_array * units.eV`?

If so, wouldn't that be rather wasteful? Say we have the potential on 10M grid points. So far, that's essentially a contiguous array of 10M `float64`s. You'd replace that with 10M pointers to as many python objects. Adding or removing units would be O(10M) sympy operations. That seems a little overkill for the purpose of attaching a single unit (~ 1 byte of information) to the whole dataset, doesn't it?

Reply:
Looks like a nice 40'000x slowdown:
![image](https://user-images.githubusercontent.com/6342175/54859922-0dcda400-4cd0-11e9-86a2-04fda0a693c0.png)

Reply:
@jgukelberger thanks for testing! You're right about the complexity - it'll be O(n) bigger and smaller. I thought it wouldn't be so bad because we convert all the arrays with units into pure numpy arrays before doing solves on them (which is the part that takes the most time). But I guess it turned out to be a lot slower :/

I opted for this b/c it's the most natural way to introduce units with our existing sympy units library (without having to make the user use a new class or library like pint https://pint.readthedocs.io/en/latest/).

Actually I was curious and speedtested pint as well, and that also has a 40x slowdown. The thin wrapper I wrote introduced almost no slowdown. I guess I'll switch back to that then
![Annotation 2019-03-22 185910](https://user-images.githubusercontent.com/4095626/54860275-00ff7f00-4cd5-11e9-8a76-31d502935ef2.jpg)